# Drug Safety Test

The dataset comes from a randomized controlled drug trial conducted by a medical group
and shared by Vanderbilt University Department of Biostatistics. The study evaluates drug
safety by comparing a Drug and Placebo group while tracking adverse effects, vital signs,
and lab measures .You will conduct hypothesis testing using t-tests to determine statistical
differences between groups.
Columns Explanation:

- age: Age of the participant
- sex: Gender of the participant (male or female).
- trx: Treatment group:
    - "Drug" → Received the actual drug.
    - "Placebo" → Received a placebo (control group).
- week: Week number in the study.
- wbc: White blood cell count (WBC) measurement.
- rbc: Red blood cell count (RBC) measurement.
- adverse_effects: Presence of adverse effects (Yes or No).
- num_effects: Number of adverse effects experienced by the participant.

> We are going to check if there is a significant difference between drug group and placebo group statics using `t-test`

## Dependencies

In [ ]:
#%pip install pandas
#%pip install scipy

In [3]:
import pandas as pd
from scipy.stats import ttest_ind

## Part 1

First we load data set and check the information of it (shape of data set, columns and their type)

In [4]:
df = pd.read_csv("drug_safety.csv")

df.head()

,age,sex,trx,week,wbc,rbc,adverse_effects,num_effects
0,62,male,Drug,0,7.3,5.1,No,0
1,62,male,Drug,1,NaN,NaN,No,0
2,62,male,Drug,12,5.6,5.0,No,0
3,62,male,Drug,16,NaN,NaN,No,0
4,62,male,Drug,2,6.6,5.1,No,0


In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16103 entries, 0 to 16102
Data columns (total 8 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   age              16103 non-null  int64  
 1   sex              16103 non-null  object 
 2   trx              16103 non-null  object 
 3   week             16103 non-null  int64  
 4   wbc              9128 non-null   float64
 5   rbc              9127 non-null   float64
 6   adverse_effects  16103 non-null  object 
 7   num_effects      16103 non-null  int64  
dtypes: float64(2), int64(3), object(3)
memory usage: 1006.6+ KB


## Part 2

we remove thos rows that contain `Nan` value with pandas built-in method `dropna`. as you can see the number of rows decrease to `9127` from `16103`. It means we lost `6976` data row but now our data is clean and we can process on it.

In [6]:
print(df.shape)

df = df.dropna()

print(df.shape)

(16103, 8)
(9127, 8)


## Part 3

Here is sample statics for numerical columns of our data set.

In [7]:
df.describe()

,age,week,wbc,rbc,num_effects
count,9127.000000,9127.000000,9127.000000,9127.000000,9127.000000
mean,64.007670,4.985428,7.340331,4.672784,0.100800
std,8.847711,4.375397,1.996645,0.458520,0.322178
min,39.000000,0.000000,1.800000,2.100000,0.000000
25%,58.000000,2.000000,6.000000,4.400000,0.000000
50%,65.000000,4.000000,7.100000,4.700000,0.000000
75%,71.000000,8.000000,8.400000,5.000000,0.000000
max,84.000000,20.000000,26.500000,7.600000,3.000000


## Part 4
We group data base on `trx` type and then we show simple statics for wanted columns.

In [13]:
trx_grouped = df.groupby("trx")
trx_statistics = trx_grouped[["wbc", "rbc", "num_effects"]].describe()
print(trx_statistics.to_string())

            wbc                                                         rbc                                              num_effects                                             
          count      mean       std  min  25%   50%  75%        max   count      mean       std  min  25%  50%  75%  max       count      mean       std  min  25%  50%  75%  max
trx                                                                                                                                                                              
Drug     6011.0  7.330461  2.008941  1.8  6.0  7.00  8.4  26.500000  6011.0  4.679130  0.454990  2.7  4.4  4.7  5.0  7.5      6011.0  0.102479  0.325529  0.0  0.0  0.0  0.0  3.0
Placebo  3116.0  7.359371  1.972890  3.0  6.0  7.15  8.4  23.799999  3116.0  4.660542  0.465083  2.1  4.4  4.7  5.0  7.6      3116.0  0.097561  0.315639  0.0  0.0  0.0  0.0  3.0


## Part 5

Using map method we map `Yes` to `1` and `No` to `0` in this way we can easly define mean of `adverse_effects` because now it's numerical.

In [9]:
df["adverse_effects"] = df["adverse_effects"].map({"Yes": 1, "No": 0})

df.head()


,age,sex,trx,week,wbc,rbc,adverse_effects,num_effects
0,62,male,Drug,0,7.3,5.1,0,0
2,62,male,Drug,12,5.6,5.0,0,0
4,62,male,Drug,2,6.6,5.1,0,0
6,62,male,Drug,4,6.9,5.2,1,1
7,62,male,Drug,8,7.1,5.0,1,1


## Part 6

### First step

We filter the data set bu `trx` type and now we have two new data frame one for `Drug` and one for `placebo`.

Here is our hypotheses:
- H₀: There is no significant difference
- H₁: There is a significant difference

We are going to use independent t-test because we use this t-test for independent samples when we want to compare the means of two independent groups or samples. We want to know if there is a significant difference between these means.


In [14]:
drug_group = df[df["trx"] == "Drug"]
placebo_group = df[df["trx"] == "Placebo"]

print(drug_group.head())
print()
print(placebo_group.head())


   age   sex   trx  week  wbc  rbc  adverse_effects  num_effects
0   62  male  Drug     0  7.3  5.1                0            0
2   62  male  Drug    12  5.6  5.0                0            0
4   62  male  Drug     2  6.6  5.1                0            0
6   62  male  Drug     4  6.9  5.2                1            1
7   62  male  Drug     8  7.1  5.0                1            1

    age   sex      trx  week  wbc  rbc  adverse_effects  num_effects
32   73  male  Placebo     0  7.2  4.7                0            0
34   73  male  Placebo    12  7.8  4.7                0            0
36   73  male  Placebo     2  7.6  4.6                0            0
38   73  male  Placebo     4  7.3  4.6                0            0
39   73  male  Placebo     8  8.2  4.6                0            0


### Second Step

basically we can calculate the `t-value` of t-test with a simple formula that requires sample means and SE s the formula is shown below:

![image.png](./images/T3-1.jpg)


Then using `t-value` and `degree of freedom` we can calculate `p-value` which is usually done by tables and calculators
and then using `t-value` and `degree of freedom` we can find `p-value` (using tables and calculators)

here we use scipy built-in method `ttest_ind` that does this all. 

> The `p-value` is the probability of getting the observed results (or something more extreme) if the null hypothesis is true.
we defina a `significant level (a)` and if p-value is less than that it means we reject H₀ otherwise we can't reject H₀

in the below part you can see the result of t-test for different significant levels.

The significance level (a) is the threshold we use to decide whether a result is statistically significant. also it shows the probility of type 1 error.

In [17]:
alpha_levels = [0.05, 0.1]
metrics = ["wbc", "rbc", "num_effects", "adverse_effects"]

for metric in metrics:
    t_stat, p_value = ttest_ind(drug_group[metric], placebo_group[metric], equal_var=False)
    print(f"{metric}:") 
    print(f"t-stat = {t_stat:.4f}, p-value = {p_value:.4f}")

    for alpha in alpha_levels:
        if p_value < alpha:
            print(f"Significant difference found with alpha = {alpha} (p < {alpha})")
        else:
            print(f"No significant difference with alpha = {alpha} ({alpha} <= p)")
    print()
print()


wbc:
t-stat = -0.6597, p-value = 0.5095
No significant difference with alpha = 0.05 (0.05 <= p)
No significant difference with alpha = 0.1 (0.1 <= p)

rbc:
t-stat = 1.8239, p-value = 0.0682
No significant difference with alpha = 0.05 (0.05 <= p)
Significant difference found with alpha = 0.1 (p < 0.1)

num_effects:
t-stat = 0.6983, p-value = 0.4850
No significant difference with alpha = 0.05 (0.05 <= p)
No significant difference with alpha = 0.1 (0.1 <= p)

adverse_effects:
t-stat = 0.6047, p-value = 0.5454
No significant difference with alpha = 0.05 (0.05 <= p)
No significant difference with alpha = 0.1 (0.1 <= p)




- What is the alternative argument and what did you choose for each metric, why?

    The alternative argument in `ttest_ind()` specifies the type of hypothesis test: `two-sided` , `less`, `greater` that two sided is the defualt option means alternative hypothese is both higher and lower. For this study, we typically use `two-sided` because we want to check if there is any difference (higher or lower) between the Drug and Placebo groups for each metric.

    However, if we had a prior expectation (for example "The drug should reduce adverse effects"), we might use "less" for num_effects and adverse_effects. but we don't have such a expectation.

- What is the equal_var argument and what does it do?

    The `equal_var` argument controls whether we assume that the two groups have equal variances.

    If equal_var=True, it assumes both groups have the same variance, which makes the test more stable but may be incorrect. If equal_var=False, it does not assume equal variances which is more reliable when the variances differ.
    Since real-world medical data often has different variances between groups, we usually set equal_var=False for better accuracy. we also have a different formula for equal variance that you can read more about it in [here](https://datatab.net/tutorial/unpaired-t-test)